pkg/controller/endpointslice/reconciler.go

```diff
@@ -73,6 +78,14 @@ func (r *reconciler) reconcile(service *corev1.Service, pods []*corev1.Pod, exis
 	for _, existingSlice := range existingSlices {
 		// service no longer supports that address type, add it to deleted slices
 		if _, ok := serviceSupportedAddressesTypes[existingSlice.AddressType]; !ok {
+			if r.topologyCache != nil {
+				svcKey, err := serviceControllerKey(existingSlice)
+				if err != nil {
+					klog.Warningf("Couldn't get key to remove EndpointSlice from topology cache %+v: %v", existingSlice, err)
+				}
+				r.topologyCache.RemoveHints(svcKey, existingSlice.AddressType)
```

Comment:
should this be deleted at the same time we delete the slice from the tracker too? L117, 
```
			r.endpointSliceTracker.ExpectDeletion(sliceToDelete)
```
just when we know that the slice  was completely deleted

Reply:
This is a really good question. I _think_ I've got the edge cases covered here or will when they exist, but I could be missing something. I think there are at least 3 relevant scenarios where we need to update the cache:

* syncService is called for a Service that has topology enabled. This will result in `AddHints` being called which resets the count of endpoints allocated per zone for the Service based on the result of that function. That can result in a call to RemoveHints if no hints should be added.
* syncService is called for a Service that does not have topology enabled. This does not exist yet, but I will add it once we have a switch. We would need to consistently call RemoveHints in this case.
* syncService is called for a Service that has one less AddressType than it did previously (and corresponding orphaned slices. There's a section starting on `reconciler.go:81` that removes these from the cache.